lib/interface-types/src/encoders/wat.rs

```diff
@@ -0,0 +1,716 @@
+        )
+    }
+}
+
+/// Encode an `Export` into a string.
+impl<'input> From<&Export<'input>> for String {
+    fn from(export: &Export) -> Self {
+        format!(
+            r#"(@interface export "{name}"{inputs}{outputs})"#,
+            name = export.name,
+            inputs = input_types_to_param(&export.input_types),
+            outputs = output_types_to_result(&export.output_types),
+        )
+    }
+}
+
+/// Encode a `Type` into a string.
+impl<'input> From<&Type<'input>> for String {
+    fn from(_ty: &Type) -> Self {
+        unimplemented!()
+    }
+}
+
+/// Encode an `Import` into a string.
+impl<'input> From<&Import<'input>> for String {
+    fn from(import: &Import) -> Self {
+        format!(
+            r#"(@interface func ${namespace}_{name} (import "{namespace}" "{name}"){inputs}{outputs})"#,
+            namespace = import.namespace,
+            name = import.name,
+            inputs = input_types_to_param(&import.input_types),
+            outputs = output_types_to_result(&import.output_types),
+        )
+    }
+}
+
+/// Encode an `Adapter` into a string.
+impl<'input> From<&Adapter<'input>> for String {
+    fn from(adapter: &Adapter) -> Self {
+        match adapter {
+            Adapter::Import {
+                namespace,
+                name,
+                input_types,
+                output_types,
+                instructions,
+            } => format!(
+                r#"(@interface adapt (import "{namespace}" "{name}"){inputs}{outputs}{instructions})"#,
+                namespace = namespace,
+                name = name,
+                inputs = input_types_to_param(&input_types),
+                outputs = output_types_to_result(&output_types),
+                instructions =
+                    instructions
+                        .iter()
+                        .fold(String::new(), |mut accumulator, instruction| {
+                            accumulator.push_str("\n  ");
+                            accumulator.push_str(&String::from(instruction));
+                            accumulator
+                        }),
+            ),
+
+            Adapter::Export {
+                name,
+                input_types,
+                output_types,
+                instructions,
+            } => format!(
+                r#"(@interface adapt (export "{name}"){inputs}{outputs}{instructions})"#,
+                name = name,
+                inputs = input_types_to_param(&input_types),
+                outputs = output_types_to_result(&output_types),
+                instructions =
+                    instructions
+                        .iter()
+                        .fold(String::new(), |mut accumulator, instruction| {
+                            accumulator.push_str("\n  ");
+                            accumulator.push_str(&String::from(instruction));
+                            accumulator
+                        }),
+            ),
+
+            _ => unimplemented!(),
```

Comment:
I'm a big advocate of always putting a string in `unimplemented`/`todo` so that it's easy to find and prints a message that a user can consume

Reply:
Done!